### PR TITLE
Guava removal

### PR DESCRIPTION
**Pull request by Apanatshka**
I'm working towards removing guava as a dependency from as much stuff as possible.
First step in this repo is to move the `MultiTable` stuff from `org.metaborg.util` to `p_raffrayi`. I've also started removing calls like `Sets.newHashSet` with a direct call to the constructor, and removing dependence on guava's `Iterables` by requiring more `Collection` instead of `Iterable`.

LMK if you have any thoughts or concerns about these (planned) changes.

### CODE REVIEW

**Comment by Apanatshka** [comment]
@AZWN do you know if there is shared mutable state in the concurrent solver?
Since I'm replacing a few datastructures with ones I've written myself, it's relevant because of thread safety...

**Comment by AZWN** [comment]
Yes, there is a little. The framework itself is designed on the actor model, meaning that there is _no_ mutable state in the actual framework itself. However, there is some mutable state in the actor framework itself (package [`mb.p_raffaryi.actors.impl`](https://github.com/metaborg/nabl/tree/master/p_raffrayi/src/main/java/mb/p_raffrayi/actors/impl), especially the `Actor` and `ActorSystem` class. However, I doubt it is relevant for your refactoring, as it is implemented using `ThreadLocal`s, `java.util.concurrent.ConcurrentLinkedDeque`, atomic references, synchronization and volatile variables, none of which uses Guava.

The CHM protocol implementation does not have shared data, so its data structures can be replaced as required.

Finally, there is some shared state in [`mb.p_raffrayi.impl.Broker`](https://github.com/metaborg/nabl/blob/master/p_raffrayi/src/main/java/mb/p_raffrayi/impl/Broker.java). This is actually a performance hhack/technical dept. I don't think it uses Guava, but if it needs to be refactored, we might as well fix it immediately.

**Comment by Apanatshka** [comment]
These are all the changes, I think. I need to push some changes in other repo's to master before this can be merged, but it can be reviewed soon.

I've broken two tests:
1. `mb.p_raffrayi.IncrementalTest#testRelease_MutualDep_ChildChanged`
2. `mb.statix.scopegraph.diff.ScopeGraphDiffTest#testGraphAdditions_TargetEarlier`

File: statix.solver/src/main/java/mb/statix/spec/ARule.java [hunk 1]
```diff
@@ -227,21 +227,32 @@ public String toString(TermFormatter termToString) {
...
+            return (r1, r2) -> LeftRightOrder.compare(r1, r2).orElse(0);
         }
 
     }
 
+    public static class LeftRightOrderWithConsistentEquality {
+        public Comparator<Rule> asComparator() {
+            return (r1, r2) -> {
+                final Optional<Integer> patternComparison = LeftRightOrder.compare(r1, r2);
+                return patternComparison.filter(i -> i != 0).orElse(Integer.compare(r1.hashCode(), r2.hashCode()));
```
Comment (AZWN) [inline statix.solver/src/main/java/mb/statix/spec/ARule.java:253]: Note to self: this does not handle hash collisions correctly.

File: statix.solver/src/main/java/mb/statix/spec/ARule.java [hunk 2]
```diff
@@ -240,18 +240,19 @@ public static Optional<Integer> compare(Rule r1, Rule r2) {
...
 
     public static class LeftRightOrderWithConsistentEquality {
         public Comparator<Rule> asComparator() {
-            return (r1, r2) -> {
-                final Optional<Integer> patternComparison = LeftRightOrder.compare(r1, r2);
-                return patternComparison.filter(i -> i != 0).orElse(Integer.compare(r1.hashCode(), r2.hashCode()));
-            };
+            return LeftRightOrder.asComparator()
+                .thenComparing((r1, r2) -> Integer.compare(r1.hashCode(), r2.hashCode()))
+                // hash collision handling...
```
Comment (AZWN) [inline statix.solver/src/main/java/mb/statix/spec/ARule.java:253]: Also this version does not handle hash collisions correctly. I'm afraid the only sound way is comparing all members of `Rule` one-by-one.

File: nabl2.solver/src/main/java/mb/nabl2/config/NaBL2DebugConfig.java [hunk 3]
```diff
@@ -21,7 +19,7 @@ public enum Flag {
 
         public static EnumSet<Flag> valuesOf(Collection<String> names) {
             Set<Flag> flags = names.stream().map(String::toUpperCase).map(Flag::valueOf).collect(Collectors.toSet());
-            return Sets.newEnumSet(flags, Flag.class);
+            return EnumSet.copyOf(flags);
```
Comment (AZWN) [inline nabl2.solver/src/main/java/mb/nabl2/config/NaBL2DebugConfig.java:22]: Alternative:

`.collect(Collectors.toCollection(() -> EnumSet.noneOf(Flag.class))`

From https://stackoverflow.com/questions/35178294/java-util-stream-collectors-with-enumset-stream

File: nabl2.solver/src/main/java/mb/nabl2/solver/components/NameResolutionComponent.java [hunk 4]
```diff
@@ -60,20 +58,18 @@ public NameResolutionComponent(SolverCore core,
...
         nameResolution.addCached(solution.resolutionCache());
-        constraints.addAll(seed(solution.declProperties(), message).constraints());
-        return SeedResult.constraints(constraints);
+        return seed(solution.declProperties(), message);
     }
 
     public SeedResult seed(IProperties<Occurrence, ITerm, ITerm> solution, IMessageInfo message)
             throws InterruptedException {
-        final java.util.Set<IConstraint> constraints = Sets.newHashSet();
+        final Set.Transient<IConstraint> constraints = Set.Transient.of();
```
Comment (AZWN) [inline nabl2.solver/src/main/java/mb/nabl2/solver/components/NameResolutionComponent.java:68]: `CapsuleUtil` performs better.

File: nabl2.solver/src/main/java/mb/nabl2/solver/solvers/FixedPointSolver.java [hunk 5]
```diff
@@ -53,25 +55,25 @@ public FixedPointSolver(ICancel cancel, IProgress progress, ISolver component) {
...
             throws InterruptedException {
         final IMessages.Transient messages = Messages.Transient.of();
 
-        final Deque<IConstraint> constraints = Lists.newLinkedList(initialConstraints);
+        final Deque<IConstraint> constraints = new ArrayDeque<>(initialConstraints);
         final IndexedBag<IConstraint, CriticalEdge> criticalEdgeDelays = new IndexedBag<>(RemovalPolicy.ALL);
         final IndexedBag<IConstraint, String> relationDelays = new IndexedBag<>(RemovalPolicy.ALL);
         final IndexedBag<IConstraint, ITermVar> variableDelays = new IndexedBag<>(RemovalPolicy.ANY);
-        final List<IConstraint> unsolved = Lists.newArrayList();
+        final Set.Transient<IConstraint> unsolved = Set.Transient.of();
```
Comment (AZWN) [inline nabl2.solver/src/main/java/mb/nabl2/solver/solvers/FixedPointSolver.java:66]: `CapsuleUtil` is more performant.

File: nabl2.solver/src/main/java/mb/nabl2/solver/solvers/FixedPointSolver.java [hunk 6]
```diff
@@ -133,15 +135,15 @@ public SolveResult solve(Iterable<? extends IConstraint> initialConstraints, Ref
         //log.info("Solved {} with {} delays, {} var delays, {} critical edge delays, {} relation delays", solvedCount,
         //        unconditionalDelayCount, variableDelayCount, criticalEdgeDelayCount, relationDelayCount);
 
-        unsolved.addAll(constraints);
-        unsolved.addAll(variableDelays.values());
-        unsolved.addAll(criticalEdgeDelays.values());
-        unsolved.addAll(relationDelays.values());
+        constraints.forEach(unsolved::__insert);
```
Comment (AZWN) [inline nabl2.solver/src/main/java/mb/nabl2/solver/solvers/FixedPointSolver.java:138]: I think there is an `__insertAll`

File: nabl2.solver/src/main/java/mb/nabl2/spoofax/primitives/SG_solve_multi_initial_constraint.java [hunk 7]
```diff
@@ -52,7 +51,7 @@ public SG_solve_multi_initial_constraint() {
                     return Tuple2.of(params, C);
                 }).match(currentTerm).orElseThrow(() -> new InterpreterException("Current term is not (params, C)."));
         final ITerm params = input._1();
-        final List<IConstraint> constraints = ImmutableList.of(input._2());
+        final Set.Immutable<IConstraint> constraints = Set.Immutable.of(input._2());
```
Comment (AZWN) [inline nabl2.solver/src/main/java/mb/nabl2/spoofax/primitives/SG_solve_multi_initial_constraint.java:54]: `CapsuleUtil` is more performant.

File: nabl2.solver/src/main/java/mb/nabl2/spoofax/primitives/SG_solve_multi_unit_constraint.java [hunk 8]
```diff
@@ -45,7 +45,7 @@ public SG_solve_multi_unit_constraint() {
                 }).match(currentTerm)
                         .orElseThrow(() -> new InterpreterException("Current term is not (InitialResult, C)."));
         final MultiInitialResult initialResult = input._1();
-        final List<IConstraint> constraints = ImmutableList.of(input._2());
+        final Set.Immutable<IConstraint> constraints = Set.Immutable.of(input._2());
```
Comment (AZWN) [inline nabl2.solver/src/main/java/mb/nabl2/spoofax/primitives/SG_solve_multi_unit_constraint.java:48]: `CapsuleUtil` is more performant.

File: nabl2.solver/src/main/java/mb/nabl2/spoofax/primitives/SG_solve_single_constraint.java [hunk 9]
```diff
@@ -53,7 +53,7 @@ public SG_solve_single_constraint() {
 
 
         final ITerm constraintTerm = ConstraintTerms.specialize(strategoTerms.fromStratego(constraintSTerm));
-        final List<IConstraint> constraints = Constraints.matchConstraintOrList().map(ImmutableList::of)
+        final Set.Immutable<IConstraint> constraints = Constraints.matchConstraintOrList().map(Set.Immutable::of)
```
Comment (AZWN) [inline nabl2.solver/src/main/java/mb/nabl2/spoofax/primitives/SG_solve_single_constraint.java:56]: `CapsuleUtil` is more performant.

File: nabl2.solver/src/main/java/mb/scopegraph/pepm16/esop15/lazy/EsopNameResolution.java [hunk 10]
```diff
@@ -314,7 +313,7 @@ private EnvL<S, L, O> stageEnv_L(Set<L> L, IRelation<L> lt) {
     }
 
     private Set<L> max(IRelation<L> lt, Set<L> L) {
-        ImmutableSet.Builder<L> maxL = ImmutableSet.builder();
+        CapsuleUtil.SetBuilder<L> maxL = new CapsuleUtil.SetBuilder<>();
```
Comment (AZWN) [inline nabl2.solver/src/main/java/mb/scopegraph/pepm16/esop15/lazy/EsopNameResolution.java:316]: Why not use a transient set directly?

File: nabl2.solver/src/main/java/mb/scopegraph/pepm16/terms/ALabel.java [hunk 11]
```diff
@@ -53,16 +52,16 @@ public abstract class ALabel extends AbstractApplTerm implements ILabel, IApplTe
         }
     }
 
-    @Value.Lazy @Override public List<ITerm> getArgs() {
```
Comment (AZWN) [inline nabl2.solver/src/main/java/mb/scopegraph/pepm16/terms/ALabel.java:56]: Prefer retaining abstract `List` interface.

File: nabl2.solver/src/main/java/mb/scopegraph/pepm16/terms/ANamespace.java [hunk 12]
```diff
@@ -33,8 +32,8 @@ public abstract class ANamespace extends AbstractApplTerm implements INamespace,
         return getName().isEmpty() ? OP0 : OP1;
     }
 
-    @Value.Lazy @Override public List<ITerm> getArgs() {
-        return getName().isEmpty() ? ImmutableList.of() : ImmutableList.of((ITerm) B.newString(getName()));
+    @Value.Lazy @Override public ImList.Immutable<ITerm> getArgs() {
```
Comment (AZWN) [inline nabl2.solver/src/main/java/mb/scopegraph/pepm16/terms/ANamespace.java:35]: prefer retaining abstract `List` interface

File: nabl2.solver/src/main/java/mb/scopegraph/pepm16/terms/AOccurrence.java [hunk 13]
```diff
@@ -44,8 +43,8 @@ public abstract class AOccurrence extends AbstractApplTerm implements IOccurrenc
         return OP;
     }
 
-    @Value.Lazy @Override public List<ITerm> getArgs() {
```
Comment (AZWN) [inline nabl2.solver/src/main/java/mb/scopegraph/pepm16/terms/AOccurrence.java:47]: prefer retaining abstract `List` interface

**Review by AZWN — COMMENTED** [review]
First batch of comments. Lets coordinate who is going to make/process these changes.

**Comment by Apanatshka** [comment]
Abstract list interface: currently I cannot get Immutables to stop using guava unless I make it more specific than the abstract list.

File: p_raffrayi/src/main/java/mb/p_raffrayi/impl/ARecordedQuery.java [hunk 14]
```diff
@@ -4,6 +4,7 @@
 
 import org.immutables.serial.Serial;
 import org.immutables.value.Value;
+import org.metaborg.util.collection.CapsuleUtil;
```
Comment (AZWN) [inline p_raffrayi/src/main/java/mb/p_raffrayi/impl/ARecordedQuery.java:7]: Why do we need this import?

Comment (Apanatshka) [reply p_raffrayi/src/main/java/mb/p_raffrayi/impl/ARecordedQuery.java:7]: It's used once in the file

**Comment by Virtlink** [comment]
Be careful not to replace one dependency (Guava) with another (Capsule). Arguably, Capsule is even less battle-tested and less maintained than Guava. Even if we cannot avoid picking _some_ implementation, generally I would prefer using the Java built-in interfaces as types, to avoid making these dependencies part of the API/function signatures and painting ourselves into a corner.

**Comment by Apanatshka** [comment]
Right now we use both all over the place. The upside of Capsule is that it is a _small_ dependencies of just those collections. Guava is _huge_ and has many parts we don't use that get CVEs filed against. This refactoring was enough work already, and I'm not going to complicate the diff any more. We can have a separate discussion on how to tie ourselves less to particular dependencies, I'm open to the idea of abstracting them away as much as possible.

**Comment by Virtlink** [comment]
Sure, I just don't understand your comment:
> Abstract list interface: currently I cannot get Immutables to stop using guava unless I make it more specific than the abstract list.
Capsule's Immutables all implement the base Java interface, right?

**Comment by Apanatshka** [comment]
Oh, the problem there is that we use a annotation processor called [Immutables](https://immutables.github.io/) which by default uses guava in its generated code. This is because guava will check if one of their collections is immutable and then not do a copy, and the generated code is all about immutability so it makes sense. If you opt into using normal Java collections in the codegen, it is supposed to no longer depend on guava, but in practice this option is broken en the codegen will still (part of the time) use guava code. The only way around it (that I'm aware of) is to not use the base Java interfaces for which it knows guava implementations.

**Comment by AZWN** [comment]
I see that the annotations have the option `@Style(jdkOnly = true)`. Documentation isn't very clear, but it seems to disable Guava. Could/should we use that instead of the very specific interfaces?

**Comment by Virtlink** [comment]
Or is there a specific issue with `jdkOnly = true` that you can link to?

**Comment by AZWN** [comment]
I found some more docs that support the intuition above:
- Javadoc: https://github.com/immutables/immutables/blob/28ba97e9f3075902e279c1cde3e16be01e365afb/value-annotations/src/org/immutables/value/Value.java#L801-L812

- https://immutables.github.io/style.html says:
  > Force the generation of JDK-only implementation code, even if Guava is available in classpath.
  
  (not explicitliy tied to `jdkOnly`, but thats the only reasonable option.

**Comment by Apanatshka** [comment]
So last open issue (if my most recent change doesn't foul up any tests) is the immutables behaviour. Two things there:
1. Perhaps the annotation in the package-info doesn't reach into sub packages and that's the issue.
2. Regardless of whether we find a solution like that, by using normal JDK stuff instead of guava, code generated by immutables will do more list copying because it cannot tell if the normal JDK list that is given as input is immutable. But it will assume this and have you handle it for custom types like the `ImList.Immutable` that I'm using in this PR. So if we can get immutables to behave, it'll be a choice between a nice API and preserving the performance we had before.

**Comment by AZWN** [comment]
Last commit indeed looks like it fixes the issue, thnx!

Regarding the interface vs. performance debate: then performance is most important.

So, basically, that means I'm happy with merging this PR now.

**Review by AZWN — APPROVED** [review]

**Pull request merged**